addons/purchase/report/purchase_report.py

```diff
@@ -155,45 +159,3 @@ def _group_by(self):
                 po.id
         """
         return group_by_str
-
-    @api.model
-    def read_group(self, domain, fields, groupby, offset=0, limit=None, orderby=False, lazy=True):
-        """ This is a hack to allow us to correctly calculate the average of PO specific date values since
-            the normal report query result will duplicate PO values across its PO lines during joins and
-            lead to incorrect aggregation values.
-
-            Only the AVG operator is supported for avg_days_to_purchase.
-        """
-        avg_days_to_purchase = next((field for field in fields if re.search(r'\bavg_days_to_purchase\b', field)), False)
-
-        if avg_days_to_purchase:
-            fields.remove(avg_days_to_purchase)
-            if any(field.split(':')[1].split('(')[0] != 'avg' for field in [avg_days_to_purchase] if field):
-                raise UserError("Value: 'avg_days_to_purchase' should only be used to show an average. If you are seeing this message then it is being accessed incorrectly.")
-
-        res = []
-        if fields:
-            res = super(PurchaseReport, self).read_group(domain, fields, groupby, offset=offset, limit=limit, orderby=orderby, lazy=lazy)
-
-        if not res and avg_days_to_purchase:
-            res = [{}]
-
-        if avg_days_to_purchase:
-            query = """ SELECT AVG(days_to_purchase.po_days_to_purchase)::decimal(16,2) AS avg_days_to_purchase
-                          FROM (
-                              SELECT extract(epoch from age(po.date_approve,po.create_date))/(24*60*60) AS po_days_to_purchase
-                              FROM purchase_order po
-                              WHERE po.id IN (
-                                  SELECT "purchase_report"."order_id" FROM %s WHERE %s)
-                              ) AS days_to_purchase
-                    """
-
-            subdomain = domain + [('company_id', '=', self.env.company.id), ('date_approve', '!=', False)]
-            subtables, subwhere, subparams = expression(subdomain, self).query.get_sql()
-
-            self.env.cr.execute(query % (subtables, subwhere), subparams)
-            res[0].update({
-                '__count': 1,
-                avg_days_to_purchase.split(':')[0]: self.env.cr.fetchall()[0][0],
-            })
-        return res
```

Comment:
I think you may have missed the point of this hack. Because of the SQL joins the days_to_purchase value gets duplicated for each `purchase.order.line` within a `purchase.order`. This means that purchase orders with more than 1 purchase order line will skew the average result. The easiest way to see this is to do the following:
- start with an empty purchase_order table
- create and confirm a new purchase order with 2 purchase order lines (with 2 different products)
- manually set the `create_date` of your new purchase order to 1 day earlier (`UPDATE purchase_order SET create_date = create_date - interval '1 day';`)
- create and confirm another new purchase order with 1 purchase order line (with a 3rd different product)
- look at the resulting "Average Days to Purchase" value in the report. It should be 0.5 days [= (1 day + 0 day) / 2 purchase orders], but it's 0.67 days [=(1 day + 1 day + 0 day)/3 purchase order lines] instead.

The same issue occurs with the sales and puchase_stock report hacks. Let me know if you need any further explanation.

Reply:
Thanks for the clear explanation :+1:.
My change is indeed wrong in that respect, I'll take it into consideration :).